MAINT: Use np.concatenate instead of np.vstack

ieee8023 opened the pull request:
The code for vstack says it is "supported for backward compatibility" so numpy itself should not use this function in other functions.

Here I changed function calls from vstack to concatenate with axis=0 to remove references to the vstack function.

My motivation is that I went to push something to Theano which does not have a vstack function and will not implement it because the vstack function in numpy is deprecated. I think it would help when porting the code in the future if this was changed over to only used supported functions.

ieee8023 committed:
MAINT: Replace internal uses of vstack

vstack is supported only for backward compatibility

We should use concatenate or stack instead

ieee8023 commented:
These tests failed on TestPiecewise which should not be related to the code I changed.

charris commented on numpy/lib/function_base.py:
```diff
@@ -2965,7 +2965,7 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
     >>> x = [-2.1, -1,  4.3]
     >>> y = [3,  1.1,  0.12]
-    >>> X = np.vstack((x,y))
+    >>> X = np.concatenate((x,y), axis=0)
```
Not the same for 1D arrays, they need to be reshaped, which is what `stack` does.

eric-wieser replied:
@ieee8023: Don't forget this one

ieee8023 replied:
I don't understand the issue here. Do you mean if the input array has shape (x,)? That doesn't work as input to vstack. Can you give me some unit tests that will fail?

With (x,1) shapes it works as follows:

```
data = np.asarray(np.random.rand(2,1))
y = np.asarray(np.random.rand(4,1))

print np.vstack([data, y]).shape
print np.concatenate([data, y], axis=0).shape

print np.allclose(np.vstack([data, y]),np.concatenate([data, y], axis=0))
```
Output:
```
(6, 1)
(6, 1)
True
```

eric-wieser replied:
> I don't understand the issue here.

The issue is that if I run your code, I get a different result to what was here before:
```python
>>> x = [-2.1, -1,  4.3]
>>> y = [3,  1.1,  0.12]
>>> np.vstack((x,y))   # before
array([[-2.1 , -1.  ,  4.3 ],
       [ 3.  ,  1.1 ,  0.12]])
>>> np.concatenate((x,y), axis=0)  # yours (wrong)
array([-2.1 , -1.  ,  4.3 ,  3.  ,  1.1 ,  0.12])
>>> np.stack((x,y), axis=0)
array([[-2.1 , -1.  ,  4.3 ],  # charris (correct)
       [ 3.  ,  1.1 ,  0.12]])
```

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1324,7 +1324,7 @@ def piecewise(x, condlist, funclist, *args, **kw):
         # Only able to stack vertically if the array is 1d or less
         if x.ndim <= 1:
-            condlist = np.vstack([condlist, ~totlist])
+            condlist = np.concatenate([condlist, ~totlist], axis=0)
```
Looks to me like this should actually be `np.stack`?

ieee8023 commented:
I don't see how np.stack can be equivalent. It requires the inputs to be the exact same shape where concatenate correctly combines the matrices.

```data = np.asarray(np.random.rand(2,5))
y = np.asarray(np.random.rand(4,5))

print np.vstack([data, y]).shape
print np.concatenate([data, y], axis=0).shape
print np.stack([data, y], axis=0).shape
```

Output is 

```(6, 5)
(6, 5)
ValueError: all input arrays must have the same shape
```

eric-wieser commented:
Right, but in this case the axes actually _do_ align.

I think this should be
```
totlist = np.any(condlist, axis=0, keepdims=True)
condlist = np.concatenate([condlist, ~totlist], axis=0)
```

eric-wieser committed:
MAINT: Remove 1d special casing in piecewise

eric-wieser commented:
Not quite what I mean - I've added an extra commit that ought to fix the tests

ieee8023 commented:
Type in the commit message: fox=fix

I added unit tests for the 1d and 2d functionality. I added a patch in the docstring so that code will work but the function already handles 1d shapes using the lines `X = array(m, ndmin=2, dtype=dtype)` and `y = array(y, copy=False, ndmin=2, dtype=dtype)` so I didn't think it made sense to add it there.

Is this the best approach? Should I use np.atleast_2d(m) in the function cov as well?

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -2965,7 +2957,7 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
     >>> x = [-2.1, -1,  4.3]
     >>> y = [3,  1.1,  0.12]
-    >>> X = np.vstack((x,y))
+    >>> X = np.concatenate([np.atleast_2d(m) for m in (x,y)], axis=0)
```
Why not just use `stack` here like we suggested?

ieee8023 replied:
I was trying to write general code. stack works in this case for 1d arrays and then concatenate is needed for 2d arrays. I changed it to stack though in 50514ed

eric-wieser replied:
Ah, I see. But this is the docs for `cov`, which is not aiming to show you how to stack matrices and vectors, so I think simplicity (`stack`) is best here.

As an aside, `np.atleast_2d(x, y)` is a thing that saves a list comprehension

eric-wieser commented on numpy/ma/tests/test_extras.py:
```diff
@@ -1123,6 +1123,16 @@ def test_2d_with_missing(self):
+
+        data = np.asarray(np.random.rand(2,))
+        y = np.asarray(np.random.rand(2,))
+        assert_almost_equal(cov(data,y=y), np.cov(data,y=y))
```
These tests don't belong here - this file is for testing the maskedarray implementation.

There is already a test for 1d data at `np.lib.tests.test_function_base.TestCov.test_1D_rowvar`, so this isn't needed at all

ieee8023 replied:
Ok, I removed them in 50514ed

eric-wieser commented:
Looks good, but could do with your commits being squashed together (with `git rebase -i`), and the commit messages amending to start with `MAINT:`.

If you're not up to doing that, I'll do so at some point

ieee8023 commented:
I am not sure I did that correctly. Should rebase-ing show all the commits like that?

eric-wieser commented:
It looks like you did a merge followed by a rebase, or vice versa.

I'd `git reset --hard 50514ed4902f127359d98a271e9c2ab2dc3f753e` and try again.

eric-wieser commented:
@ieee8023: Alright, I've rebased this for you, but now someone else should probably review it

charris commented on numpy/ma/tests/test_extras.py:
```diff
@@ -1122,7 +1122,6 @@ def test_2d_with_missing(self):
                             (np.cov(xf, rowvar=False, bias=True) *
                              x.shape[0] / frac))
 
-
```
The blank line is correct PEP8, please put it back.

charris committed:
STY: Fix missing blank line in ma.tests.test_extras.py

charris commented:
I fixed it, will squash merge when tests finish.

charris commented:
@eric-wieser How did you rebase it?

charris commented:
Thanks @ieee8023 .
